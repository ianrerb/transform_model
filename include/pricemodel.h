#ifndef PRICEMODEL_H
#define PRICEMODEL_H
#include<complex.h>
#include<math.h>
#include<vector>

/* pricemodel is an abstract base class (ABC).  
 * User cannot declare variables of type pricemodel
 * pricemodel provides interface for derived classes GBM, [...]
 */

class pricemodel {
  public:
    pricemodel(double interest_rate=0.0, double div_rate=0.0, double time=0.0) : r(interest_rate), q(div_rate), T(time){ };
    virtual ~pricemodel(){ };

    void setTimeToExpiry(double &time) { T=time; };
    void setRiskFreeRate(double &rate) { r=rate; };
    void setDividendRate(double &div_rate) { q=div_rate; };

    double TimeToExpiry() const {return T; };
    double RiskFreeRate() const {return r; };
    double DividendRate() const {return q; };
    
    //computes E[exp(iulogS)]
    virtual double complex logCF(const double &S, const double complex &u) const=0;
  
  protected:
    double r;
    double q;
    double T;
  };

class GBM : public pricemodel {
  public:
    GBM(double volatility, double interest_rate=0.0, double div_rate=0.0, double time=0.0): pricemodel(interest_rate, div_rate, time), sigma(volatility){ };
    virtual ~GBM(){ };
    
    void setVolatility(double &vol) {sigma = vol; };
    double Volatility() const {return sigma; };

    double complex logCF(const double &S, const double complex &u) const; //defined in GBM.cpp    
  private:
    double sigma;
  };

class Heston : public pricemodel {
  public:
    Heston(double kappa_, double theta_, double sigma_, double init_vol, double rho_=0.0,  double interest_rate=0.0, double div_rate=0.0, double time=0.0): pricemodel(interest_rate, div_rate, time),kappa(kappa_), theta(theta_), sigma(sigma_), v0(init_vol), rho(rho_) {}; 
    ~Heston(){};

    void setKappa(double &kappa_) {kappa = kappa_; };
    void setTheta(double &theta_) {theta = theta_; };
    void setSigma(double &sigma_) {sigma = sigma_; };
    void setRho(double &rho_) {rho = rho_; };
    void setInitialVolatility(double &vol) {v0=vol;};

    double Kappa() const {return kappa; };
    double Theta() const {return theta; };
    double Sigma() const {return sigma; };
    double Rho() const {return rho; };
    double InitialVolatility() const {return v0; };

    double complex logCF(const double &S, const double complex &u) const; //defined in Heston.cpp    

  private:
    double kappa, theta, sigma, rho, v0;
  };
#endif
