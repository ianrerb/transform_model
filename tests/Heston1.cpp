#include"../include/pricemodel.h"
#include<cmath>
#include<iostream>

using namespace std;
void printComplex(double complex c){
  cout<<"("<<creal(c)<<","<<cimag(c)<<")"<<endl;  
  }

int main(){
  Heston model(2.0,.04,.5,.04,-0.7,.03,0.0,.5);
  double complex val;
  val = model.logCF(1,1); 
  printComplex(val);
   
  val = model.logCF(100,0);
  printComplex(val);
  
  val = model.logCF(exp(1),1); 
  printComplex(val);
  return 0;
  }
