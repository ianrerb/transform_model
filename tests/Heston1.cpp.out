(5.12095,1.53107)
(5.36556,0)
(5.12095,1.53107)
